File: tests/test_project_clients.mjs
// Real app UI with an isolated mock API. No email, invitations or project edits outside the fixture.
// PLAYWRIGHT_MODULE=/path/to/playwright CHROMIUM_PATH=/path/to/chrome node --test tests/test_project_clients.mjs
import assert from 'node:assert/strict';
import {before,after,test} from 'node:test';
import {createServer} from 'node:http';
import {readFile,mkdir} from 'node:fs/promises';
import {createRequire} from 'node:module';
import {fileURLToPath} from 'node:url';
import {resolve,extname} from 'node:path';
import {demoRequest} from '../public/assets/demo.js';

const {chromium}=createRequire(import.meta.url)(process.env.PLAYWRIGHT_MODULE||'playwright');
const root=fileURLToPath(new URL('../public/',import.meta.url));
const template=await demoRequest('project',{id:'van-galen'});
let server,browser,base;
before(async()=>{
 server=createServer(async(req,res)=>{
  try{
   const path=new URL(req.url,'http://localhost').pathname;
   const file=path.startsWith('/assets/')?resolve(root,'.'+path):resolve(root,'index.html');
   if(!file.startsWith(root)){res.writeHead(404);res.end();return;}
   res.setHeader('Content-Type',({'.html':'text/html','.js':'text/javascript','.css':'text/css','.svg':'image/svg+xml','.webp':'image/webp'})[extname(file)]||'application/octet-stream');
   res.end(await readFile(file));
  }catch{res.writeHead(404);res.end();}
 });
 await new Promise((done,reject)=>{server.once('error',reject);server.listen(0,'127.0.0.1',done);});
 base=`http://127.0.0.1:${server.address().port}`;
 browser=await chromium.launch({headless:true,args:['--no-sandbox'],...(process.env.CHROMIUM_PATH?{executablePath:process.env.CHROMIUM_PATH}:{})});
});
after(async()=>{await browser?.close();if(server)await new Promise(done=>server.close(done));});

async function setup(t,{width=1440,empty=false,editable=true,rejectShare=false}={}){
 const page=await browser.newPage({viewport:{width,height:950}});page.setDefaultTimeout(6000);
 const errors=[],calls=[];page.on('pageerror',e=>errors.push(e.message));
 t.after(async()=>{await page.close();assert.deepEqual(errors,[]);});
 const studio={id:'client-test-studio',name:'Test studio',role:'member'};
 const deck={...structuredClone(template),project:{...template.project,id:'client-test-project',name:'Client selection',studio_id:studio.id},
  iteration:{...template.iteration,id:'client-test-iteration',project_id:'client-test-project',number:1,status:'draft'},
  clients:empty?[]:[{name:'Alice <A>',email:'alice@example.test'},{name:'Bob',email:'bob@example.test'}],
  can_edit:editable,members:[],files:[],slides:[],budget:[],comments:[],events:[],jobs:[],shares:[],contacts:[],
  slide_content:[],slide_layout:[],slide_sections:[],slide_groups:{story:'Story'},capabilities:{ai:false,mail:false}};
 deck.iterations=[deck.iteration];
 await page.route('**/*',async route=>{
  const url=new URL(route.request().url());
  if(url.origin!==base){await route.abort();return;}
  if(url.pathname!=='/api.php'){await route.continue();return;}
  const action=url.searchParams.get('action'),data=route.request().postDataJSON();calls.push({action,data});
  let result,status=200;
  if(action==='session')result={user:{id:'editor',name:'Editor',email:'editor@example.test'},csrf:'test-csrf',studio,studios:[studio],capabilities:deck.capabilities};
  else if(action==='projects')result={projects:[{...deck.project,iteration:deck.iteration,can_edit:editable}]};
  else if(action==='project')result=deck;
  else if(action==='share'){
   if(rejectShare){status=409;result={error:'A selected client is no longer a member of this project. Refresh and choose again.'};}
   else {deck.iteration.status='shared';result={links:data.client_emails.map(email=>({id:email,email,sent:false,url:base+'/#/view/test-token'}))};}
  }else if(action==='save_project_client'){
   const old=deck.clients.find(c=>c.email===data.email);if(old)old.name=data.name;else deck.clients.push({name:data.name,email:data.email});result={clients:deck.clients};
  }else if(action==='remove_project_client'){deck.clients=deck.clients.filter(c=>c.email!==data.email);result={clients:deck.clients};}
  else if(action==='drive_status')result={configured:false,connected:false};
  else {errors.push('Unexpected API call: '+action);status=500;result={error:'Unexpected API call'};}
  await route.fulfill({status,contentType:'application/json',body:JSON.stringify(result)});
 });
 await page.goto(`${base}/${studio.id}/projects/${deck.project.id}`);
 await page.locator('.project-clients-inline').waitFor();
 return {page,calls,deck};
}

for(const width of [1440,390])test(`All clients default selected; sends only checked recipients at ${width}px`,async t=>{
 const {page,calls}=await setup(t,{width});
 await page.locator('.project-head [data-action=share]').click();
 const form=page.locator('[data-form=share]');await form.waitFor();
 assert.equal(await form.locator('[name=client_email]:checked').count(),2);
 assert.equal(await form.locator('textarea[name=emails]').count(),0);
 assert.equal(await form.locator('a').count(),0,'Client names are escaped');
 const message=form.locator('[name=message]');await message.fill('Please review this iteration.');
 await form.getByRole('button',{name:'Clear selection',exact:true}).click();
 assert.ok(await form.locator('[type=submit]').isDisabled());
 await form.getByRole('button',{name:'Select all',exact:true}).click();
 assert.equal(await form.locator('[name=client_email]:checked').count(),2);
 await form.locator('[value="bob@example.test"]').uncheck();
 assert.ok(await form.locator('[type=submit]').isEnabled());
 assert.equal(await message.inputValue(),'Please review this iteration.');
 assert.ok(await page.getByRole('dialog').evaluate(el=>el.scrollWidth<=el.clientWidth+1));
 if(process.env.TEST_SCREENSHOT_DIR){
  await mkdir(process.env.TEST_SCREENSHOT_DIR,{recursive:true});
  await page.screenshot({path:resolve(process.env.TEST_SCREENSHOT_DIR,`client-selection-${width}.png`)});
 }
 await form.locator('[type=submit]').click();
 await page.locator('.link-result').waitFor();
 assert.equal(await page.locator('.link-result').count(),1);
 assert.deepEqual(calls.find(c=>c.action==='share').data,{iteration:'client-test-iteration',client_emails:['alice@example.test'],message:'Please review this iteration.'});
 await page.getByRole('button',{name:'Done',exact:true}).click();
 if(process.env.TEST_SCREENSHOT_DIR)await page.screenshot({path:resolve(process.env.TEST_SCREENSHOT_DIR,`project-actions-${width}.png`)});
 await page.locator('.project-head [data-action=share]').click();
 assert.equal(await page.locator('[name=client_email]:checked').count(),2,'Every new send starts with the whole client list selected');
});

test('Client management is independent of sending and supports add, edit and removal',async t=>{
 const {page,calls}=await setup(t,{empty:true});
 await page.locator('.project-head [data-action=share]').click();
 assert.match(await page.getByRole('dialog').innerText(),/Add your clients first/);
 await page.getByRole('dialog').getByRole('button',{name:'Manage clients',exact:true}).click();
 await page.getByRole('button',{name:'Add client',exact:true}).click();
 let form=page.locator('[data-form=project-client]');
 await form.locator('[name=name]').fill('Chris');await form.locator('[name=email]').fill('chris@example.test');
 await form.locator('[type=submit]').click();await page.locator('.project-client-row').waitFor();
 assert.equal(calls.filter(c=>c.action==='share').length,0);
 await page.getByRole('button',{name:'Edit Chris',exact:true}).click();
 form=page.locator('[data-form=project-client]');assert.ok(await form.locator('[name=email]').evaluate(el=>el.readOnly));
 await form.locator('[name=name]').fill('Chris Updated');await form.locator('[type=submit]').click();
 await page.getByRole('button',{name:'Remove Chris Updated',exact:true}).click();
 const removal=page.locator('[data-form=remove-project-client]');await removal.waitFor();
 assert.match(await page.getByRole('dialog').innerText(),/access to all its shared iterations will end/);
 await removal.getByRole('button',{name:'Cancel',exact:true}).click();
 assert.equal(calls.filter(c=>c.action==='remove_project_client').length,0);
 await page.locator('[data-action=project-clients]').click();
 await page.getByRole('button',{name:'Remove Chris Updated',exact:true}).click();
 await page.locator('[data-form=remove-project-client] [type=submit]').click();
 await page.getByRole('dialog').getByText('No client members yet.',{exact:false}).waitFor();
 assert.deepEqual(calls.find(c=>c.action==='remove_project_client').data,{project_id:'client-test-project',email:'chris@example.test'});
});

test('Read-only viewers cannot manage clients or send invitations',async t=>{
 const {page}=await setup(t,{editable:false});
 assert.equal(await page.locator('[data-action=project-clients],[data-action=share]').count(),0);
 assert.match(await page.locator('.project-clients-inline').innerText(),/Alice <A>/);
});

test('A stale selection shows the server error and preserves the message',async t=>{
 const {page}=await setup(t,{rejectShare:true});
 await page.locator('.project-head [data-action=share]').click();
 const form=page.locator('[data-form=share]');await form.locator('[name=message]').fill('Keep this message');
 await form.locator('[type=submit]').click();await form.locator('.form-error').waitFor();
 assert.match(await form.locator('.form-error').innerText(),/no longer a member/);
 assert.equal(await form.locator('[name=message]').inputValue(),'Keep this message');
 assert.equal(await form.locator('[name=client_email]:checked').count(),2);
});

File: tests/test_project_clients.py
"""Project client rosters and per-iteration recipient access, with isolated data.
Run: PHP_BIN=php python3 tests/test_project_clients.py -v
"""
import json
import unittest

from test_security import Client, SecurityFixture, png


class ProjectClientTests(SecurityFixture):
    def add(self, email, name='Client', project='own', actor=None):
        return self.ok((actor or self.editor).api('save_project_client',
                       {'project_id': project, 'email': email, 'name': name}))['clients']

    def send(self, emails, iteration='iteration-own'):
        return self.ok(self.editor.api('share', {'iteration': iteration, 'client_emails': emails}))['links']

    def account(self, email):
        client = Client(self.base)
        self.ok(client.api('request_login', {'email': email}))
        token = (self.tmp / 'mail.log').read_text().strip().split('/#/login/')[-1]
        self.ok(client.api('consume_login', {'token': token}))
        return client

    def test_project_client_list_is_independent_of_team_and_invitations(self):
        members = self.sql('SELECT * FROM project_members ORDER BY project_id')
        shares = self.sql('SELECT * FROM shares ORDER BY id')
        users = self.sql('SELECT * FROM users ORDER BY id')
        clients = self.add('Alice@Example.Test', 'Alice')
        self.assertEqual([(c['name'], c['email']) for c in clients], [('Alice', 'alice@example.test')])
        self.add('alice@example.test', 'Alice Updated')
        self.assertEqual(self.sql('SELECT name FROM project_client_members WHERE project_id=?', ('own',)), [('Alice Updated',)])
        self.assertEqual(self.sql('SELECT * FROM project_members ORDER BY project_id'), members)
        self.assertEqual(self.sql('SELECT * FROM shares ORDER BY id'), shares)
        self.assertEqual(self.sql('SELECT * FROM users ORDER BY id'), users)
        self.assertFalse((self.tmp / 'mail.log.messages.jsonl').exists())
        self.assertEqual(self.ok(self.editor.api('project', query={'id': 'own'}))['clients'][0]['name'], 'Alice Updated')
        self.assertNotIn('clients', self.ok(self.client.api('deck')))
        p = self.ok(self.editor.api('create_project', {'name': 'With clients', 'emails': ['a@example.test', 'b@example.test']}), 201)
        self.assertEqual({c['email'] for c in self.ok(self.editor.api('project', query={'id': p['project_id']}))['clients']},
                         {'a@example.test', 'b@example.test'})

    def test_only_project_team_can_manage_clients(self):
        self.add('alice@example.test', 'Alice')
        for actor, project in ((self.anon, 'own'), (self.client, 'own'), (self.admin, 'own'),
                               (self.editor, 'private'), (self.editor, 'public'), (self.editor, 'foreign')):
            for action in ('save_project_client', 'remove_project_client'):
                with self.subTest(actor=actor.session, project=project, action=action):
                    self.denied(actor.api(action, {'project_id': project, 'email': 'alice@example.test', 'name': 'Changed'}))
        self.denied(self.editor.api('save_project_client', {'project_id': 'own', 'email': 'x@example.test', 'name': 'X'},
                                    headers={'X-CSRF-Token': ''}))
        self.assertEqual(self.editor.api('save_project_client', {'project_id': 'own', 'email': "' OR 1=1--", 'name': 'X'})[0], 400)
        self.assertEqual(self.sql('SELECT name FROM project_client_members WHERE project_id=?', ('own',)), [('Alice',)])

    def test_selected_clients_alone_get_each_iteration(self):
        self.add('alice@example.test', 'Alice')
        self.add('bob@example.test', 'Bob')
        alice_link = self.send(['alice@example.test'])[0]
        self.assertEqual(self.sql('SELECT email FROM shares WHERE iteration_id=?', ('iteration-own',)), [('alice@example.test',)])
        alice, bob = self.account('alice@example.test'), self.account('bob@example.test')
        alice.client_share = alice_link['id']
        bob.client_share = alice_link['id']  # A guessed selector cannot grant access.
        self.assertEqual(self.ok(alice.api('deck'))['iteration']['id'], 'iteration-own')
        self.assertEqual(alice.api('file', query={'id': 'file-own'})[1], png('file-own'))
        self.denied(bob.api('deck'))
        self.denied(bob.api('file', query={'id': 'file-own'}))
        self.assertEqual(self.ok(bob.api('destinations'))['projects'], [])
        second = self.ok(self.editor.api('new_iteration', {'iteration': 'iteration-own'}), 201)['id']
        bob_link = self.send(['bob@example.test'], second)[0]
        bob.client_share = bob_link['id']
        self.assertEqual(self.ok(bob.api('deck'))['iteration']['id'], second)
        self.denied(alice.api('deck', query={'iteration': second}))
        self.denied(bob.api('deck', query={'iteration': 'iteration-own'}))
        self.assertEqual(self.ok(alice.api('destinations'))['projects'][0]['iteration_id'], 'iteration-own')
        self.assertEqual({c['email'] for c in self.ok(self.editor.api('project', query={'id': 'own'}))['clients']},
                         {'alice@example.test', 'bob@example.test'})
        self.send(['alice@example.test'], second)
        # Deselecting Bob on this later send does not revoke his earlier invitation.
        self.ok(bob.api('deck'))
        messages = [json.loads(line) for line in (self.tmp / 'mail.log.messages.jsonl').read_text().splitlines()]
        self.assertEqual([m['to'] for m in messages], ['alice@example.test', 'bob@example.test', 'alice@example.test'])

    def test_invalid_or_stale_recipient_selection_does_not_publish(self):
        self.add('alice@example.test', 'Alice')
        self.add('outside@example.test', 'Outside', 'private', Client(self.base, 'colleague'))
        for selection, status in (([], 400), ('alice@example.test', 400),
                                  (['outside@example.test'], 409),
                                  (['alice@example.test', 'missing@example.test'], 409),
                                  (['alice@example.test'] * 21, 400)):
            with self.subTest(selection=selection):
                response = self.editor.api('share', {'iteration': 'iteration-own', 'client_emails': selection})
                self.assertEqual(response[0], status, response[1])
                self.assertEqual(self.sql('SELECT status FROM iterations WHERE id=?', ('iteration-own',)), [('draft',)])
                self.assertEqual(self.sql('SELECT COUNT(*) FROM shares WHERE iteration_id=?', ('iteration-own',)), [(0,)])
        self.ok(self.editor.api('remove_project_client', {'project_id': 'own', 'email': 'alice@example.test'}))
        self.assertEqual(self.editor.api('share', {'iteration': 'iteration-own', 'client_emails': ['alice@example.test']})[0], 409)
        self.assertFalse((self.tmp / 'mail.log.messages.jsonl').exists())

    def test_removal_revokes_all_project_grants_aliases_and_queued_notifications(self):
        self.add('alice@example.test', 'Alice')
        self.add('bob@example.test', 'Bob')
        links = self.send(['alice@example.test', 'bob@example.test'])
        self.ok(self.editor.api('comment', {'iteration': 'iteration-own', 'body': 'Review this design', 'slide': 'intro'}))
        aliases = self.sql('SELECT url FROM email_outbox WHERE email=?', ('alice@example.test',))
        self.assertEqual(len(aliases), 1)
        other = self.ok(self.editor.api('create_project', {'name': 'Another project', 'emails': ['alice@example.test'], 'starting_pack': []}), 201)
        other_link = self.send(['alice@example.test'], other['iteration_id'])[0]
        alice = self.account('alice@example.test')
        self.ok(self.editor.api('remove_project_client', {'project_id': 'own', 'email': 'alice@example.test'}))
        alice.client_share = links[0]['id']
        self.denied(alice.api('deck'))
        bearer = Client(self.base, bearer=links[0]['id'])
        self.denied(bearer.api('file', query={'id': 'file-own'}))
        alias = Client(self.base, bearer=aliases[0][0])
        self.denied(alias.api('deck'))
        self.assertEqual(self.sql('SELECT status FROM email_outbox WHERE email=?', ('alice@example.test',)), [('cancelled',)])
        alice.client_share = other_link['id']
        self.ok(alice.api('deck'))
        self.assertEqual({p['id'] for p in self.ok(alice.api('destinations'))['projects']}, {other['project_id']})
        bob = self.account('bob@example.test');bob.client_share = links[1]['id']
        self.ok(bob.api('deck'))
        self.add('alice@example.test', 'Alice Again')
        alice.client_share = links[0]['id']
        self.denied(alice.api('deck'))  # Re-adding membership must not revive revoked invitations.
        self.assertEqual(self.sql('PRAGMA foreign_key_check'), [])

    def test_existing_client_contacts_and_share_recipients_migrate_once(self):
        self.sql('DELETE FROM project_client_members')
        self.sql("INSERT INTO contacts(id,project_id,name,role,email) VALUES('legacy-client','own','Legacy Alice','Client','Alice@Example.Test')")
        self.sql("DELETE FROM migrations WHERE name='project-clients-v1'")
        own = self.ok(self.editor.api('project', query={'id': 'own'}))
        shared = self.ok(self.editor.api('project', query={'id': 'shared'}))
        self.assertEqual([(c['email'], c['name']) for c in own['clients']], [('alice@example.test', 'Legacy Alice')])
        self.assertEqual([c['email'] for c in shared['clients']], ['client@example.test'])
        self.ok(self.editor.api('remove_project_client', {'project_id': 'shared', 'email': 'client@example.test'}))
        self.assertEqual(self.ok(self.editor.api('project', query={'id': 'shared'}))['clients'], [])
        self.assertEqual(self.ok(self.editor.api('project', query={'id': 'shared'}))['clients'], [])
        self.denied(self.client.api('deck'))

    def test_legacy_email_invites_and_client_contacts_join_the_roster(self):
        self.ok(self.editor.api('share', {'iteration': 'iteration-own', 'emails': ['legacy@example.test']}))
        self.ok(self.editor.api('save_contact', {'project_id': 'own', 'role': 'Client', 'email': 'contact@example.test', 'name': 'Contact'}))
        self.assertEqual({c['email'] for c in self.ok(self.editor.api('project', query={'id': 'own'}))['clients']},
                         {'legacy@example.test', 'contact@example.test'})


if __name__ == '__main__':
    unittest.main()
